src/main/scala/io/citrine/lolo/linear/LinearRegression.scala

```diff
@@ -26,82 +29,86 @@ case class LinearRegressionLearner(
       trainingData: Seq[(Vector[Any], Any)],
       weights: Option[Seq[Double]]
   ): LinearRegressionTrainingResult = {
-    val n = trainingData.size
+    val lambda = regParam.getOrElse(0.0)
+    val regularized = lambda > 0.0
+
+    val (features, labels) = trainingData.unzip
+    val rep = features.head
 
     /* Get the indices of the continuous features */
-    val indices: Vector[Int] = trainingData.head._1.zipWithIndex
-      .filter(_._1.isInstanceOf[Double])
-      .filterNot(_._1.asInstanceOf[Double].isNaN)
-      .map(_._2)
-      .filterNot(i => trainingData.exists(_._1(i).asInstanceOf[Double].isNaN))
-      .filterNot { i =>
-        val unregularized = !regParam.exists(_.asInstanceOf[Double] > 0.0)
-        lazy val constant = trainingData.forall(_._1(i) == trainingData.head._1(i))
-        unregularized && constant // remove constant features if there's no regularization
+    val indices = rep.zipWithIndex
+      .collect { case (x: Double, idx) if !x.isNaN => idx }
```

Comment:
Is this redundant? It's removing features for which the representative row has a value of NaN. But then the next block removes features for which any row has a value of NaN.

Reply:
At first I was leaving it in because I thought it would save us a list traversal by pre-filtering. But, I actually think it causes an extra traversal over the `rep` because `hasNans.exists` will exit early on satisfaction, and `constant` is lazy. 

So yeah, its redundant and should be removed.

Reply:
Ehh, the filter/collect logic turns out a bit cleaner by pre-collecting the `Double` indices, so I'm going to leave as is.